### PR TITLE
Extend InputTypeMixin

### PR DESCRIPTION
**Pull request by radka-j**
The motivation for this PR is primarily to set us up for #371.

Specifically, this PR:
- moves the `InputTypeMixin` class from `experimental.emulators.base.py` to `experimental.data.py`
- renames `_convert()` to `_convert_to_dataloader()`
- adds `_convert_to_dataset()`
- adds `_random_split()`
- extends the `Tuner` to subclass `InputTypeMixin` 

It also does a bit of docstrings clean up in `experimental/` to make it consistent with the rest of the repository.

Questions:
- I am still a bit unsure where the input data should be kept as a dataset vs. loader. I suspect some of this will become clearer as we also implement `Autoemulate.compare()`.

### CODE REVIEW

File: autoemulate/experimental/data.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        elif isinstance(x, DataLoader) and y is not None:
+            raise ValueError(
+                f"Since x is already a DataLoader, expect y to be None, not {type(y)}."
+            )
+        else:
+            dataset = self._convert_to_dataset(x, y)
+            dataloader = DataLoader(dataset, batch_size=batch_size, shuffle=shuffle)
+        return dataloader
+
+    def _random_split(
```
Comment (radka-j) [inline autoemulate/experimental/data.py:59]: After talking to @sgreenbury, we agreed that `_random_split` should return a `Dataset` not `DataLoader`

File: autoemulate/experimental/data.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+        Parameters
+        ----------
+        dataset: Dataset
+            The data to split.
+        batch_size: int | None
+            The DataLoader batch_size. If None, sets batch_size to lenth of training
+            data. Defaults to None.
+        """
+        train, test = tuple(random_split(dataset, [0.8, 0.2]))
```
Comment (sgreenbury) [inline autoemulate/experimental/data.py:73]: Perhaps we can have the train size as a parameter in method

File: autoemulate/experimental/data.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            )
+
+        return dataset
+
+    def _convert_to_dataloader(
+        self,
+        x: InputLike,
+        y: OutputLike | None = None,
+        batch_size: int = 16,
+        shuffle: bool = True,
```
Comment (radka-j) [inline autoemulate/experimental/data.py:43]: We should also make sure that the higher level classes that call this method include a shuffle argument to pass to it.

File: autoemulate/experimental/data.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+    def _convert_to_dataset(
+        self,
+        x: InputLike,
+        y: OutputLike | None = None,
+    ) -> Dataset:
+        """
+        Convert input data to pytorch Dataset.
+        """
+        # Convert input to DataLoader if not already
```
Comment (radka-j) [inline autoemulate/experimental/data.py:21]:
```suggestion
        # Convert input to Dataset if not already
```

File: autoemulate/experimental/data.py (hunk 1 above)
Comment (sgreenbury) [reply autoemulate/experimental/data.py]: As discussed, can leave as `tuple[DataLoader, DataLoader]` here but to be refactored in #371

File: autoemulate/experimental/tuner.py [hunk 5]
```diff
@@ -24,20 +31,18 @@ class Tuner:
         The validation scores and parameter values used in each search iteration.
     """
 
-    def __init__(self, dataset: Dataset, n_iter: int):
+    def __init__(self, x: InputLike, y: OutputLike | None, n_iter: int):
```
Comment (EdwinB12) [inline autoemulate/experimental/tuner.py:34]: Should y be InputLike here insteadof OutputLike?

Comment (EdwinB12) [inline autoemulate/experimental/tuner.py:34]:
```suggestion
    def __init__(self, x: InputLike, y: InputLike | None, n_iter: int):
```

**Comment by EdwinB12** [comment]
Sorry I was a bit late to this but some nice changes in here! Nice work!

File: autoemulate/experimental/tuner.py (hunk 5 above)
Comment (sgreenbury) [reply autoemulate/experimental/tuner.py:34]: Thanks - yes agreed!
